boards: st: stm32h573i_dk: Add board variant for app in external flash

tpambor opened the pull request:
Commit https://github.com/zephyrproject-rtos/zephyr/commit/5d0306c797b35b7a701071f18114dc32062a9be6 of https://github.com/zephyrproject-rtos/zephyr/pull/88579 moved the `slot0_partition`, `slot1_partition`, `scratch_partition` from the internal flash to the external OSPI NOR flash. At the same time, the `nor` partition on the external NOR flash was deleted. This breaks all applications that rely on the `slot0_partition`, `slot1_partition`, `scratch_partition` to be on internal flash as well as all applications that use the `nor` partition on external flash with the `stm32h573i_dk` board.

To avoid breaking existing projects and maintain compatibility with previous versions of Zephyr, this PR restores the default flash layout (without scratch partition) and introduces a new board variant `ext_flash_app` for applications stored in external flash. This approach allows users to opt-in to the new flash layout without affecting existing projects.

A similar to what is proposed for the STM32H750B-DK in #97037.

Fixes #97423.

tpambor commented:
> @tpambor Thanks for this, but don't you miss to add an overlay in mcuboot repo to provide the right `zephyr,flash` and `zephyr,flash-controller` ? (Dependency that I'd like to get rid of but mandatory today for current version of #97037)

You are right, I missed that, currently it places mcuboot in external flash, so a overlay is necessary for mcuboot.
```
	/* node '/chosen' defined in zephyr/dts/common/skeleton.dtsi:13 */
	chosen {
...
		zephyr,flash = &ext_flash;                 /* in zephyr/boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app.dts:15 */
		zephyr,flash-controller = &ext_flash_ctrl; /* in zephyr/boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app.dts:16 */
		zephyr,code-partition = &boot_partition;   /* in bootloader/mcuboot/boot/zephyr/app.overlay:3 */
	};
```

etienne-lms reviewed (commented):
The overall looks good to me and is more consistent regarding the board initial layout.

For commit "boards: st: stm32h573i_dk: Add board variant for app in external flash":
Would it be worth to add an entry in the board doc to mention the new variant?

For commit "mgmt: hawkbit: Replace overlay with new board variant":
Could you prefix the commit message header line with `samples: `?

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,62 @@
+			reg = <0x01000000 DT_SIZE_M(16)>;
+		};
+
+		scratch_partition: partition@2000000 {
```
Scratch partition is not needed.

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk-common.dtsi:
```diff
@@ -0,0 +1,409 @@
+		reg = <0x90000000 DT_SIZE_M(64)>;
+		zephyr,memory-region = "EXTMEM";
+		/* The ATTR_MPU_EXTMEM attribut causing a MPU FAULT */
+		zephyr,memory-attr = <DT_MEM_ARM(ATTR_MPU_IO)>;
```
```suggestion
		zephyr,memory-attr = <DT_MEM_ARM_MPU_FLASH>;
```

JarmouniA replied:
Weird, can you verify if `CONFIG_MPU_ALLOW_FLASH_WRITE` is enabled in MCUboot's generate `.config`?

JarmouniA replied:
Weird, can you verify if `CONFIG_MPU_ALLOW_FLASH_WRITE` is enabled in MCUboot's generate `.config`?
On STM32H750B-DK, `MPU_IO` worked fine for basic apps, but with LVGL the app wouldn't start, I had to disable the MPU. With `MPU_FLASH`, everything works.

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk-common.dtsi:
```diff
@@ -0,0 +1,409 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		ext_flash: mx25lm51245: ext-flash@90000000 {
```
```suggestion
		ext_flash: mx25lm51245: ext-flash@0 {
```

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk.dts:
```diff
@@ -308,6 +30,21 @@
+			reg = <0x00078000 DT_SIZE_K(416)>;
+		};
+
+		scratch_partition: partition@e0000 {
```
swap-using-scratch mode is deprecated and will be removed.
https://docs.mcuboot.com/design.html#image-swap-using-scratch

JarmouniA reviewed (commented):
Add `Kconfig.sysbuild` with:
```
choice BOOTLOADER
	default BOOTLOADER_MCUBOOT
endchoice

choice BOOT_SIGNATURE_TYPE
	default BOOT_SIGNATURE_TYPE_NONE
endchoice

if BOARD_STM32H573I_DK_STM32H573XX_EXT_FLASH_APP

choice MCUBOOT_MODE
	default MCUBOOT_MODE_DIRECT_XIP
endchoice

endif # BOARD_STM32H573I_DK_STM32H573XX_EXT_FLASH_APP
```

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,57 @@
+		/* Set 64KB of storage at the end of Bank1 */
+		storage_partition: partition@f0000 {
+			label = "storage";
+			reg = <0x000f0000 DT_SIZE_K(64)>;
```
And give storage all the remaining flash.

tpambor commented on boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,62 @@
+
+		slot0_partition: partition@0 {
+			label = "image-0";
+			reg = <0x00000000 DT_SIZE_M(16)>;
```
I have removed the scratch partition, switched to swap mode and increased the two slots.

JarmouniA commented on soc/st/stm32/Kconfig.defconfig:
```diff
@@ -55,6 +55,11 @@ DT_FLASH_PARENT_IS_XSPI := $(dt_node_has_compat,$(DT_CHOSEN_FLASH_PARENT),$(DT_C
+DT_CHOSEN_Z_FLASH_CTRL := zephyr,flash-controller
+DT_COMPAT_XSPI_CTRL := st,stm32-xspi-nor
+DT_CHOSEN_FLASH_CTRL_NODE := $(dt_chosen_path,$(DT_CHOSEN_Z_FLASH_CTRL))
+DT_FLASH_CTRL_IS_XSPI := $(dt_node_has_compat,$(DT_CHOSEN_FLASH_CTRL_NODE),$(DT_COMPAT_XSPI_CTRL))
```
```suggestion
DT_COMPAT_XSPI_NOR_CTRL := st,stm32-xspi-nor
DT_CHOSEN_FLASH_CTRL_NODE := $(dt_chosen_path,$(DT_CHOSEN_Z_FLASH_CTRL))
DT_FLASH_CTRL_IS_XSPI_NOR := $(dt_node_has_compat,$(DT_CHOSEN_FLASH_CTRL_NODE),$(DT_COMPAT_XSPI_CTRL))
```

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk-common.dtsi:
```diff
@@ -0,0 +1,410 @@
+		reg = <0x90000000 DT_SIZE_M(64)>;
+		zephyr,memory-region = "EXT_FLASH";
+		/* The ATTR_MPU_EXTMEM attribut causing a MPU FAULT */
+		zephyr,memory-attr = <DT_MEM_ARM(ATTR_MPU_IO)>;
```
```suggestion
		/* DT_MEM_ARM_MPU_EXTMEM and DT_MEM_ARM_MPU_FLASH cause MPU issues */
		zephyr,memory-attr = <DT_MEM_ARM_MPU_IO>;
```

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app.yaml:
```diff
@@ -0,0 +1,31 @@
+  - zephyr
+sysbuild: true
+ram: 640
+flash: 32767
```
```suggestion
flash: 31744 # size in kB of 1 app slot minus MCUboot header size (1KB)
```

JarmouniA replied:
Ah sorry, silly mistake on my part.

JarmouniA reviewed (commented):
A note about MCUboot samples and tests, you should have both board variants in the `platform_allow` lists, since both can run with MCUboot.

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk.dts:
```diff
@@ -308,6 +30,16 @@
 
+		slot0_partition: partition@10000 {
+			label = "image-0";
+			reg = <0x00010000 DT_SIZE_K(448)>;
```
Last part of Etienne's [comment](https://github.com/zephyrproject-rtos/zephyr/pull/97678#discussion_r2434863784) still needs to be addressed. To be memory efficient, 1st slot has to be 1 sector larger than the 2nd.
https://docs.mcuboot.com/design.html#image-swap-no-scratch

maass-hamburg commented on boards/st/stm32h573i_dk/Kconfig.sysbuild:
```diff
@@ -0,0 +1,18 @@
+if BOARD_STM32H573I_DK_STM32H573XX_EXT_FLASH_APP
+
+choice MCUBOOT_MODE
+	default MCUBOOT_MODE_DIRECT_XIP
```
we should revert it back.
this default should not be set on a board leven, when the new default is incompatible with normal behavior.
in tree there are some uses of where MCUBOOT_MODE_OVERWRITE_ONLY is set as the default, but that wouldn't require 2 different images.

maass-hamburg replied:
> Direct XIP was already the default for this board, see https://github.com/mcu-tools/mcuboot/blob/e4fc5ae221cee2ce9f1120afc2092a693ac5e8a2/boot/zephyr/boards/stm32h573i_dk.conf#L1. This settings only moved from mcuboot repo to `Kconfig.sysbuild`, there is no functional change here.

for sysbuild this would have been overwritten by the sysbuild default.

after `west build -p --sysbuild -b stm32h573i_dk samples/subsys/mgmt/hawkbit/`
.config in mcuboot sysbuild build dir:
```

#
# MCUboot-specific configuration options
#
CONFIG_MCUBOOT=y
CONFIG_BOOT_USE_MBEDTLS=y

#
# MCUBoot settings
#
# CONFIG_SINGLE_APPLICATION_SLOT is not set
CONFIG_BOOT_SOMETHING_USES_SHA256=y
CONFIG_BOOT_IMG_HASH_ALG_SHA256_ALLOW=y
# CONFIG_BOOT_IMG_HASH_DIRECTLY_ON_STORAGE is not set
CONFIG_BOOT_IMG_HASH_ALG_SHA256=y
# CONFIG_BOOT_SIGNATURE_TYPE_NONE is not set
CONFIG_BOOT_SIGNATURE_TYPE_RSA=y
CONFIG_BOOT_SIGNATURE_TYPE_RSA_LEN=2048
# CONFIG_BOOT_SIGNATURE_TYPE_ECDSA_P256 is not set
# CONFIG_BOOT_SIGNATURE_TYPE_ED25519 is not set
# CONFIG_BOOT_BYPASS_KEY_MATCH is not set
CONFIG_BOOT_SIGNATURE_KEY_FILE="/home/fmaass@vogl.zone/zephyrproject/bootloader/mcuboot/root-rsa-2048.pem"
CONFIG_MCUBOOT_CLEANUP_ARM_CORE=y
# CONFIG_MCUBOOT_CLEANUP_RAM is not set
CONFIG_MBEDTLS_CFG_FILE="config-mbedtls.h"
# CONFIG_BOOT_HW_KEY is not set
CONFIG_BOOT_VALIDATE_SLOT0=y
# CONFIG_BOOT_PREFER_SWAP_OFFSET is not set
# CONFIG_BOOT_PREFER_SWAP_MOVE is not set
# CONFIG_BOOT_SWAP_USING_SCRATCH is not set
# CONFIG_BOOT_UPGRADE_ONLY is not set
# CONFIG_BOOT_SWAP_USING_OFFSET is not set
CONFIG_BOOT_SWAP_USING_MOVE=y
# CONFIG_BOOT_DIRECT_XIP is not set
# CONFIG_BOOT_RAM_LOAD is not set
# CONFIG_BOOT_FIRMWARE_LOADER is not set
# CONFIG_BOOT_BOOTSTRAP is not set
# CONFIG_SINGLE_APPLICATION_SLOT_RAM_LOAD is not set
CONFIG_BOOT_ENCRYPTION_SUPPORT=y
```

nordicjm replied:
I'm OK with boards using different defaults for the mode, sure some other features might not be compatible with a specific configuration but that's just how those features are, direct-xip is a valid supported MCUboot mode and fine for any board to use (by default or otherwise)

maass-hamburg commented on samples/subsys/mgmt/hawkbit/sample.yaml:
```diff
@@ -6,10 +6,10 @@ common:
   platform_allow:
     - frdm_k64f
-    - stm32h573i_dk
+    - stm32h573i_dk/stm32h573xx/ext_flash_app
```
i still prefer both variants working with hawkbit, independent if they are tested

tpambor replied:
Works with `stm32h573i_dk`. For `stm32h573i_dk/stm32h573xx/ext_flash_app` no longer compiles with #98061. Without #98061 booted fine for me, but problem is also more that it would require two images. So for the moment I removed `stm32h573i_dk/stm32h573xx/ext_flash_app`.

tpambor replied:
Works fine with `MCUBOOT_MODE_SINGLE_APP`. I added the board variant with external flash again to the hawkbit sample.

nordicjm commented on boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app_defconfig:
```diff
@@ -0,0 +1,18 @@
+# Enable HW stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+# Enable uart driver
```
*UART

tpambor replied:
Done

nordicjm commented on boards/st/stm32h573i_dk/stm32h573i_dk_stm32h573xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,62 @@
+		slot1_partition: partition@1000000 {
+			label = "image-1";
+			reg = <0x01000000 DT_SIZE_M(16)>;
+		};
```
you add it here https://github.com/zephyrproject-rtos/zephyr/pull/97678/commits/4dc5aef42cdb4282b702cd09a56732b72be0428d then remove it here https://github.com/zephyrproject-rtos/zephyr/pull/97678/commits/0fe1f40a93bfdf7c3f2749694d02f790c32c1d7c -> as below comment, squash these fixup commits

JarmouniA reviewed (commented):
Can you please rework the 2 commits concerning the board, put everything concerning the base variant (with the common dtsi) in a 1st commit which builds and runs correctly with & without MCUboot. Then, introduce the ext_flash_app variant in a 2nd commit.

JarmouniA commented on samples/subsys/mgmt/hawkbit/sample.yaml:
```diff
@@ -7,9 +7,11 @@ common:
   integration_platforms:
     - frdm_k64f
     - stm32h573i_dk
+    - stm32h573i_dk/stm32h573xx/ext_flash_app
```
```suggestion
```

JarmouniA reviewed (commented):
Single App mode does not support image upgrade.

JarmouniA commented on boards/st/stm32h573i_dk/stm32h573i_dk-common.dtsi:
```diff
@@ -0,0 +1,410 @@
+			compatible = "soc-nv-flash";
+			reg = <0x0 DT_SIZE_M(64)>;
+			write-block-size = <1>;
+			erase-block-size = <DT_SIZE_K(64)>;
```
@tpambor I switched H750B-DK to swap-using-offset after fixing the erase-block-size (which fixed all swap modes), up to you if you want to do the same here.

maass-hamburg replied:
Just do it like in the other PR, but add both variants to the hawkbit sample test.

tpambor replied:
Yes, now it boots here also in swap mode. I kept both variants for the hawkbit sample.

tpambor committed:
boards: st: stm32h573i_dk: Add board variant for app in external flash

Restore old internal flash layout as default and make app in external
flash a new board variant

Switch mcuboot to use swap mode and remove the no longer required
scratch partitions from the internal and external flash layout.

Introduce a 'soc-nv-flash' node in the stm32h573i_dk board DTS file
to define the external flash memory parameters used for MCUBoot
integration. This way BOOT_MAX_IMG_SECTORS_AUTO can be used to
automatically calculate the maximum number of image sectors based
on the actual external flash configuration.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
soc: st: stm32: Use flash-controller to detect running from external flash

Use the zephyr,flash-controller node and the respective compatible
'st,stm32-xspi-nor' to detect if the device is running from external
flash, instead of assuming that the parent node of the flash is the
SPI controller. This fixes an issue when soc-nv-flash is used as the
zephyr,flash node.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
samples: sysbuild: with_mcuboot: Replace overlay with new board variant

Replace overlay for stm32h573i_dk with new board variant designed to
run in external flash. Keep stm32h573i_dk as it also can run mcuboot.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
samples: mgmt: hawkbit: Replace overlay with new board variant

Replace overlay for stm32h573i_dk with new board variant designed to
run from external flash.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
tests: boot: test_mcuboot: add stm32h573i_dk

Add stm32h573i_dk and the ext_flash_app board to the allowed boards
for the mcuboot tests.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

djiatsaf-st commented on tests/boot/test_mcuboot/testcase.yaml:
```diff
@@ -63,6 +65,8 @@ tests:
     platform_allow:
       - b_u585i_iot02a
       - stm32h750b_dk/stm32h750xx/ext_flash_app
+      - stm32h573i_dk
```
ditto
